### PR TITLE
Fix/rhel7 domain join

### LINKED ISSUE #1661: CentOS 7: joining domain is not working properly. Realmd?
From the GUI, joining the domain deson't finish properly. (loading line fills and then nothing happens)

/usr/local/pf/logs/httpd.admin.log-20160817.gz:Aug 16 11:20:13 httpd.admin(30994) INFO: Memory configuration is not valid anymore for key config::Domain in local cached_hash (pfconfig::cached::is_valid)
/usr/local/pf/logs/httpd.admin.log-20160817.gz:Aug 16 11:22:20 httpd.admin(30994) INFO: domain join : Failed to join domain: failed to lookup DC info for domain 'inverse.local' over rpc: {Device Timeout} The specified I/O operation on %hs was not completed before the time-out period expired.
/usr/local/pf/logs/httpd.admin.log-20160817.gz:Aug 16 11:23:45 httpd.admin(30994) INFO: Hard expiring resource : config::Domain (pfconfig::manager::expire)
/usr/local/pf/logs/httpd.admin.log-20160817.gz:Aug 16 11:23:45 httpd.admin(30994) INFO: Expiring child resource resource::domain_dns_servers. Master resource is config::Domain (pfconfig::manager::expire)
/usr/local/pf/logs/httpd.admin.log-20160817.gz:Aug 16 11:23:45 httpd.admin(30994) INFO: Hard expiring resource : resource::domain_dns_servers (pfconfig::manager::expire)
/usr/local/pf/logs/httpd.admin.log-20160817.gz:Aug 16 11:21:14 httpd.admin(21876) INFO: Memory configuration is not valid anymore for key config::Domain in local cached_hash (pfconfig::cached::is_valid)
/usr/local/pf/logs/httpd.admin.log-20160817.gz:Aug 16 11:23:21 httpd.admin(21876) INFO: domain join : Failed to join domain: failed to lookup DC info for domain 'inverse.local' over rpc: The remote network is not reachable by the transport.

/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:11 pfcmd.pl(31264) INFO: Expiring child resource resource::domain_dns_servers. Master resource is config::Domain (pfconfig::manager::expire)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:11 pfcmd.pl(31264) INFO: Hard expiring resource : resource::domain_dns_servers (pfconfig::manager::expire)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:11 pfcmd.pl(31262) INFO: Memory configuration is not valid anymore for key config::Domain (pfconfig::manager::is_valid)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:11 pfcmd.pl(31264) INFO: Memory configuration is not valid anymore for key config::Domain in local cached_hash (pfconfig::cached::is_valid)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:11 pfcmd.pl(31262) INFO: Memory configuration is not valid anymore for key config::Domain in local cached_hash (pfconfig::cached::is_valid)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:12 pfcmd.pl(31262) WARN: Problem trying to run command: LANG=C sudo ip netns add INVERSE called from manager::winbindd_child::build_namespaces. Child exited with non-zero value 1 (pf::util::pf_run)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:12 pfcmd.pl(31262) WARN: Problem trying to run command: LANG=C sudo ip link add INVERSE-a type veth peer name INVERSE-b called from manager::winbindd_child::build_namespaces. Child exited with non-zero value 2 (pf::util::pf_run)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:12 pfcmd.pl(31262) WARN: Problem trying to run command: LANG=C sudo ip link set INVERSE-a netns INVERSE called from manager::winbindd_child::build_namespaces. Child exited with non-zero value 1 (pf::util::pf_run)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:12 pfcmd.pl(31262) WARN: Problem trying to run command: LANG=C sudo ip netns exec INVERSE route add default gw 169.254.0.2 dev INVERSE-a called from manager::winbindd_child::build_namespaces. Child exited with non-zero value 7 (pf::util::pf_run)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:12 pfcmd.pl(31264) INFO: Daemon winbindd-INVERSE.conf took 0.106 seconds to start. (pf::services::manager::launchService)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:12 pfcmd.pl(31262) INFO: Daemon winbindd-INVERSE.conf took 0.117 seconds to start. (pf::services::manager::launchService)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:46 httpd.portal(31306) INFO: [mac:[undef]] Memory configuration is not valid anymore for key config::Pf in local cached_hash (pfconfig::cached::is_valid)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:46 httpd.portal(31306) INFO: [mac:[undef]] Instantiate a new iptables modification method. pf::ipset (pf::inline::get_technique)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:48 httpd.portal(29909) INFO: [mac:[undef]] Instantiate a new iptables modification method. pf::ipset (pf::inline::get_technique)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:20:48 httpd.portal(29909) INFO: [mac:[undef]] Instantiate a new iptables modification method. pf::ipset (pf::inline::get_technique)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:21:12 pfcmd.pl(31264) WARN: winbindd-INVERSE.conf timed out trying to start (pf::services::manager::postStartCleanup)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:26 pfcmd.pl(31511) INFO: pidof -x winbindd-INVERSE.conf returned 31451 (pf::services::manager::pidFromFile)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:26 pfcmd.pl(31511) INFO: verifying process 31451 (pf::services::manager::removeStalePid)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:26 pfcmd.pl(31511) INFO: pidof -x winbindd-INVERSE.conf returned 31451 (pf::services::manager::pidFromFile)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:27 pfcmd.pl(31511) WARN: Problem trying to run command: LANG=C sudo kill -0 31451 >/dev/null 2>&1 called from (eval). Child exited with non-zero value 1 (pf::util::pf_run)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:27 pfcmd.pl(31511) INFO: removing stale pid file /usr/local/pf/var/run/INVERSE/winbindd.pid (pf::services::manager::removeStalePid)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:27 pfcmd.pl(31511) INFO: pidof -x httpd.admin returned 21875 (pf::services::manager::pidFromFile)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:27 pfcmd.pl(31511) INFO: verifying process 21875 (pf::services::manager::removeStalePid)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:27 pfcmd.pl(31511) INFO: pidof -x httpd.admin returned 21875 (pf::services::manager::pidFromFile)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:27 pfcmd.pl(31511) INFO: pidof -x httpd.admin returned 21875 (pf::services::manager::pidFromFile)
/usr/local/pf/logs/packetfence.log-20160817.gz:Aug 16 11:22:27 pfcmd.pl(31511) INFO: Instantiate profile default (pf::Portal::ProfileFactory::_from_profile)

Joining domain with realm suite worked out of the box (not in chroot though)
yum install realmd
realm join INVERSE.local -U administrator --client-software=winbind

Should we go in that direction?

### PR DESCRIPTION
**Pull request by tlaurion**
# Description

-Fixes domain join issues encountered with CentOS7 linked to passing to version changes of samba tighter ip namespace implementation of newer kernels
-Domain join progress bar now returns before 300 seconds, thanks to inotify code now checking subtree for pid creation where it should (instead of timeout for active process check for file never being created). 

Fixes:
-Path to Samba credential cache is not overwritten anymore every time winbind is relaunched, resulting in a a possible testjoin. 
-Using chroot inside a ip netns invalidates netns automatic usage of netns's resolv.conf. bind mounting each first level directory under /etc, and copying over first level files under /etc over to be chrooted directory.
-Leaving the domain wasn't done in the chroot.
-Winbind pid inotify was watching for file creation in the wrong directory,resulting in superlong winbind restarts (timeout, after which process existence was actively validated.)
-Domain join was made with "-S IP DOMAIN_NAME". Joining domain with -S required PDC name, and fails otherwise. The code for join and leave don't supply it anymore. How was it working with prior version?
-Debian8 pid creation is the same behavior that CentoOS7. Code patched for
# Impacts

-None. Tested against CentOS6/7, Debian7/8.
# Code / PR Dependencies

-pid watch is now relative to OS validated pid path.
# Issue

fixes #1661 
# Delete branch after merge

YES
## Bug Fixes
- Issue with joining multiple domains with CentOS7  (#1661)
- Fixes CentOS6 domain join that actually overwrote credential files each time create_chroot was called.
- Domain join progression bar fix for Centos7/Debian8.

### CODE REVIEW

File: lib/pf/domain.pm [hunk 1]
```diff
@@ -132,12 +136,17 @@ Joins the domain through the ip namespace
...
+    
+    use Data::Dumper;
+    print Dumper("unjoin_domain: info=" . $info);
+    print Dumper("unjoin_domain: domain=" . $domain);
+
     if($info){
-        my ($status, $output) = run("/usr/bin/sudo /sbin/ip netns exec $domain net ads leave -S $info->{ad_server} $info->{dns_name} -s /etc/samba/$domain.conf -U '$info->{bind_dn}%$info->{bind_pass}'");
+        my ($status, $output) = run("/usr/bin/sudo /sbin/ip netns exec $domain /usr/sbin/chroot $chroot_path net ads leave -s /etc/samba/$domain.conf -U '$info->{bind_dn}%$info->{bind_pass}'");
         $logger->info("domain leave : ".$output);
-        $logger->info("netns deletion : ".run("/usr/bin/sudo /sbin/ip netns delete $domain"));
```
Comment (tlaurion) [inline lib/pf/domain.pm]: Oups. deleted that. Testing revert and pushing changes.

File: addons/create_chroot.sh [hunk 2]
```diff
@@ -7,23 +7,25 @@ if [ -z "$NS" ] || [ -z "$BASE" ]; then
...
+DIRS=(etc proc lib lib64 bin usr sbin sys var/lib/samba dev var/log/samba)
 
 MOUNTS=(`mount | awk '{print $3}'`)
 
 for dir in "${DIRS[@]}"; do
   value=$BASE/$NS/$dir
   if [[ ! " ${MOUNTS[@]} " =~ " ${value} " ]]; then
-    mount -o bind /$dir           $BASE/$NS/$dir
-  fi
+    if [[ "$dir" != "etc" ]]; then
```
Comment (julsemaan) [inline addons/create_chroot.sh]: tabulation is wrong for this whole block

File: lib/pf/domain.pm [hunk 3]
```diff
@@ -93,9 +92,14 @@ sub join_domain {
     my $chroot_path = chroot_path($domain);
 
     regenerate_configuration();
-
+    
     my $info = $ConfigDomain{$domain};
-    my ($status, $output) = run("/usr/bin/sudo /sbin/ip netns exec $domain /usr/sbin/chroot $chroot_path net ads join -S $info->{ad_server} $info->{dns_name} -s /etc/samba/$domain.conf -U '$info->{bind_dn}%$info->{bind_pass}'");
+    
+    use Data::Dumper;
```
Comment (julsemaan) [inline lib/pf/domain.pm]: Remove those

File: lib/pf/domain.pm [hunk 4]
```diff
@@ -132,10 +136,16 @@ Joins the domain through the ip namespace
 sub unjoin_domain {
     my ($domain) = @_;
     my $logger = get_logger();
+    my $chroot_path = chroot_path($domain);
 
     my $info = $ConfigDomain{$domain};
+    
+    use Data::Dumper;
```
Comment (julsemaan) [inline lib/pf/domain.pm]: Remove those

File: lib/pf/services/manager/winbindd_child.pm [hunk 5]
```diff
@@ -114,6 +115,34 @@ sub pidFile {
     }
 }
 
+=head2 _setupWatchForPidCreate
+
+This setups a watch on the run directory to wait for the pid to
```
Comment (julsemaan) [inline lib/pf/services/manager/winbindd_child.pm]: *This setups a watch on the run directory and its childs to wait for the pid to appear

File: addons/create_chroot.sh [hunk 6]
```diff
@@ -7,23 +7,25 @@ if [ -z "$NS" ] || [ -z "$BASE" ]; then
     exit 1;
 fi
 
-DIRS=(proc var etc lib lib64 usr sbin bin var/cache sys var/lib/samba dev tmp run/samba var/log/samba)
+DIRS=(proc var etc lib lib64 usr sbin bin var/cache/samba sys var/lib/samba dev tmp run/samba var/log/samba var/lock var/run)
```
Comment (julsemaan) [inline addons/create_chroot.sh]: Attempt removing var/lock and var/run to see if all still succeeds as I believe the issue you fixed isn't tied to that

**Review by julsemaan — CHANGES_REQUESTED** [review]

File: lib/pf/domain.pm (hunk 3 above)
Comment (tlaurion) [inline lib/pf/domain.pm]: oupsi

**Comment by tlaurion** [comment]
Current state should work on all CentOS7.

File: lib/pf/services/manager/winbindd_child.pm [hunk 7]
```diff
@@ -75,12 +75,6 @@ sub build_namespaces(){
         my $OUTERLOGDIRECTORY="$CHROOT_PATH/$LOGDIRECTORY";
         my $OUTERRUNDIRECTORY="$CHROOT_PATH/var/run/samba$domain";
         my $PIDDIRECTORY="$var_dir/run/$domain";
-        my $DOMAIN_IF="$domain-b";
```
Comment (tlaurion) [inline lib/pf/services/manager/winbindd_child.pm]: That is not working in debian.

**Review by julsemaan — APPROVED** [review]

**Comment by julsemaan** [comment]
Changing milestone to 6.3 as this can't be cherry-picked into maintenance as it changes the samba cache directory which would make working installations break

**Pull request merged**